pf4j/src/main/java/org/pf4j/PluginClassLoader.java

```diff
@@ -107,135 +112,95 @@ public void addFile(File file) {
                 log.trace("Found loaded class '{}'", className);
                 return loadedClass;
             }
-
-            if (!parentFirst) {
-                // nope, try to load locally
+            for (ClassLoadingSource classLoadingSource : classLoadingOrder) {
+                Class<?> c = null;
                 try {
-                    loadedClass = findClass(className);
-                    log.trace("Found class '{}' in plugin classpath", className);
-                    return loadedClass;
-                } catch (ClassNotFoundException e) {
-                    // try next step
-                }
-
-                // look in dependencies
-                loadedClass = loadClassFromDependencies(className);
-                if (loadedClass != null) {
-                    log.trace("Found class '{}' in dependencies", className);
-                    return loadedClass;
-                }
-
-                log.trace("Couldn't find class '{}' in plugin classpath. Delegating to parent", className);
+                    switch (classLoadingSource) {
+                        case PARENT:
+                            c = super.loadClass(className);
+                            break;
+                        case PLUGIN:
+                            c = findClass(className);
+                            break;
+                        case DEPENDENCIES:
+                            c = loadClassFromDependencies(className);
+                            break;
+                    }
+                } catch (ClassNotFoundException ignored) {
 
-                // use the standard ClassLoader (which follows normal parent delegation)
-                return super.loadClass(className);
-            } else {
-                // try to load from parent
-                try {
-                    return super.loadClass(className);
-                } catch (ClassCastException e) {
-                    // try next step
-                }
-
-                log.trace("Couldn't find class '{}' in parent. Delegating to plugin classpath", className);
-
-                // nope, try to load locally
-                try {
-                    loadedClass = findClass(className);
-                    log.trace("Found class '{}' in plugin classpath", className);
-                    return loadedClass;
-                } catch (ClassNotFoundException e) {
-                    // try next step
                 }
-
-                // look in dependencies
-                loadedClass = loadClassFromDependencies(className);
-                if (loadedClass != null) {
-                    log.trace("Found class '{}' in dependencies", className);
-                    return loadedClass;
+                if (c != null) {
+                    log.trace("Found class '{}' in " + classLoadingSource + " classpath", className);
```

Comment:
Please use `log.trace("Found class '{}' in {} classpath", className, classLoadingSource);`. It's short and readable.
Also see for other `log.X` methods changed by this PR.

Reply:
done